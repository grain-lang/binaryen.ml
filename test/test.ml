open Binaryen

(* Testing colors enable *)
let _ = assert (Settings.are_colors_enabled () == true)
let _ = Settings.set_colors_enabled false
let _ = assert (Settings.are_colors_enabled () == false)
let _ = Settings.set_colors_enabled true

(* Testing debug_info enable *)
let _ = assert (Settings.get_debug_info () == false)
let _ = Settings.set_debug_info true
let _ = assert (Settings.get_debug_info () == true)
let _ = Settings.set_debug_info false

(* Testing low_memory_unused enable *)
let _ = assert (Settings.get_low_memory_unused () == false)
let _ = Settings.set_low_memory_unused true
let _ = assert (Settings.get_low_memory_unused () == true)
let _ = Settings.set_low_memory_unused false
let wasm_mod = Module.create ()
let _ = Module.set_features wasm_mod [ Module.Feature.all ]
let import_wasm_mod = Module.create ()

let _ =
  Import.add_memory_import import_wasm_mod "internal_name" "external_name"
    "external_base_name" true

let _ =
  assert (Import.memory_import_get_module import_wasm_mod = "external_name")

let _ =
  assert (Import.memory_import_get_base import_wasm_mod = "external_base_name")

let _ = assert (Memory.is_shared import_wasm_mod = true)

(* Testing Return.get_value *)
let _ =
  assert (
    Option.is_some
      (Expression.Return.get_value
         (Expression.Return.make wasm_mod
            (Expression.Const.make wasm_mod (Literal.int32 0l)))))

let _ =
  assert (
    Option.is_none
      (Expression.Return.get_value
         (Expression.Return.make wasm_mod (Expression.Null.make ()))))

(* Create function type for i32 (i32, i32) *)
let params () = Type.create [| Type.int32; Type.int32 |]
let results = Type.int32
let x () = Expression.Local_get.make wasm_mod 0 Type.int32
let y () = Expression.Local_get.make wasm_mod 1 Type.int32
let load = Expression.Load.make wasm_mod 1 ~signed:true 0 0 Type.int32 (y ())

let select =
  Expression.Select.make wasm_mod
    (Expression.Const.make wasm_mod (Literal.int32 1l))
    (x ()) load

let bin = Expression.Binary.make wasm_mod Op.add_int32 select (y ())

let if_ =
  Expression.If.make wasm_mod
    (Expression.Const.make wasm_mod (Literal.int32 0l))
    (Expression.Unreachable.make wasm_mod)
    (Expression.Null.make ())

let _ = assert (Expression.If.get_if_false if_ = None)

let add =
  Expression.Block.make wasm_mod ~return_type:Type.int32 "add" [ if_; bin ]

let _ = assert (Expression.Block.get_name add = Some "add")

(* Create the add function *)
let adder = Function.add_function wasm_mod "adder" (params ()) results [||] add

let call_adder =
  Expression.Call_indirect.make wasm_mod "table"
    (Expression.Const.make wasm_mod (Literal.int32 0l))
    [
      Expression.Const.make wasm_mod (Literal.int32 3l);
      Expression.Const.make wasm_mod (Literal.int32 5l);
    ]
    (params ()) Type.int32

let start =
  Function.add_function wasm_mod "start" Type.none Type.none [||]
    (Expression.Drop.make wasm_mod call_adder)

let _ = Export.add_function_export wasm_mod "adder" "adder"
let _ = Table.add_table wasm_mod "table" 1 1 Type.funcref
let funcref_expr1 = Expression.Ref.func wasm_mod "adder" Type.funcref

let _ =
  Expression.Table.set wasm_mod "table"
    (Expression.Const.make wasm_mod (Literal.int32 0l))
    funcref_expr1

let funcref_expr2 =
  Expression.Table.get wasm_mod "table"
    (Expression.Const.make wasm_mod (Literal.int32 0l))
    Type.funcref

let _ = Expression.print funcref_expr2
let table_size = Expression.Table.size wasm_mod "table"
let _ = Expression.print table_size
let table_name = Expression.Table_size.get_table table_size
let _ = Expression.Table_size.set_table table_size table_name
let null_ref = Expression.Ref.null wasm_mod Type.funcref

let table_grow =
  Expression.Table.grow wasm_mod "table" null_ref
    (Expression.Const.make wasm_mod (Literal.int32 0l))

let _ = Expression.print table_grow

let _ =
  Global.add_global wasm_mod "max_int64" Type.int64 false
    (Expression.Const.make wasm_mod (Literal.int64 Int64.max_int))

let _ =
  Global.add_global wasm_mod "max_int64_mut" Type.int64 true
    (Expression.Const.make wasm_mod (Literal.int64 Int64.max_int))

let _ =
  Global.add_global wasm_mod "test_float64_bits" Type.float64 false
    (Expression.Const.make wasm_mod (Literal.float64_bits 0x3FF3AE147AE147AEL))

let _ =
  Table.add_active_element_segment wasm_mod "table" "elem" [ "adder" ]
    (Expression.Const.make wasm_mod (Literal.int32 0l))

let _ = Function.set_start wasm_mod start

let segment : Binaryen.Memory.segment =
  let data = Bytes.of_string "hello" in
  let kind =
    Binaryen.Memory.Active
      { offset = Expression.Const.make wasm_mod (Literal.int32 0l) }
  in
  let size = Bytes.length data in
  { data; kind; size }

let passive_segment : Binaryen.Memory.segment =
  let data = Bytes.of_string "world" in
  let kind = Binaryen.Memory.Passive in
  let size = Bytes.length data in
  { data; kind; size }

let _ = assert (Memory.has_memory wasm_mod = false)

let _ =
  Memory.set_memory wasm_mod 1 Memory.unlimited "memory"
    [ segment; passive_segment ]
    false

let _ = assert (Memory.has_memory wasm_mod = true)
let _ = assert (Memory.get_initial wasm_mod = 1)
let _ = assert (Memory.has_max wasm_mod = false)
let _ = assert (Memory.get_max wasm_mod = Memory.unlimited)
let max_memory_wasm_mod = Module.create ()
let _ = Memory.set_memory max_memory_wasm_mod 1 2 "memory" [] false
let _ = assert (Memory.has_max max_memory_wasm_mod = true)
let _ = assert (Memory.get_max max_memory_wasm_mod = 2)

(* Create an imported "write" function i32 (anyref, i32, i32) *)
(* Similar to the example here: https://bytecodealliance.org/articles/reference-types-in-wasmtime *)

let _ =
  Import.add_function_import wasm_mod "write" "future-wasi" "write"
    (Type.create [| Type.anyref; Type.int32; Type.int32 |])
    Type.int32

(* Create a function that calls the imported write function *)
let _ =
  Function.add_function wasm_mod "hello" Type.anyref Type.int32 [||]
    (Expression.Call.make wasm_mod "write"
       [
         Expression.Local_get.make wasm_mod 0 Type.anyref;
         Expression.Const.make wasm_mod (Literal.int32 0l);
         Expression.Const.make wasm_mod (Literal.int32 1l);
       ]
       Type.int32)

let _ = Export.add_function_export wasm_mod "hello" "hello"

(* Shouldn't actually do anything since we aren't doing function renames *)
let _ = Module.update_maps wasm_mod

(* Finally, we print 3 versions of the module to be checked against test.expected *)

(* 1. Print the the module as-is *)

let _ = Module.print wasm_mod

(* 2. Optimize, then print the module *)

let _ = Module.optimize wasm_mod
let _ = Module.print wasm_mod

(* 3. Copy previous module bytes into new module, validate, and print *)

let byts, no_sourcemap = Module.write wasm_mod None
let _ = assert (no_sourcemap = None)
let _, sourcemap = Module.write wasm_mod (Some "test.wasm.map")

let _ =
  assert (
    sourcemap = Some {|{"version":3,"sources":[],"names":[],"mappings":""}|})

let new_mod = Module.read byts

let _ =
  Module.run_passes new_mod
    [ Passes.name_types; Passes.merge_similar_functions; Passes.spill_pointers ]

let _ =
  Module.set_features new_mod
    [
      Module.Feature.mvp;
      Module.Feature.atomics;
      Module.Feature.bulk_memory;
      Module.Feature.mutable_globals;
      Module.Feature.nontrapping_fp_to_int;
      Module.Feature.sign_ext;
      Module.Feature.simd128;
      Module.Feature.exception_handling;
      Module.Feature.tail_call;
      Module.Feature.reference_types;
      Module.Feature.multivalue;
      Module.Feature.gc;
      Module.Feature.memory64;
      Module.Feature.typed_function_references;
      Module.Feature.relaxed_simd;
      Module.Feature.extended_const;
      Module.Feature.all;
    ]

let _ = Module.validate new_mod
let _ = Module.print new_mod

let _ = Module.print_stack_ir new_mod

(* Dispose the modules ðŸ‘‹ *)

let _ = Module.dispose wasm_mod
let _ = Module.dispose new_mod
