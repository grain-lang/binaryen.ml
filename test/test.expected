(table.get $table
 (i32.const 0)
)
(table.size $table)
(table.grow $table
 (ref.null func)
 (i32.const 0)
)
(module
 (type $i32_i32_=>_i32 (func (param i32 i32) (result i32)))
 (type $none_=>_none (func))
 (type $anyref_i32_i32_=>_i32 (func (param anyref i32 i32) (result i32)))
 (type $anyref_=>_i32 (func (param anyref) (result i32)))
 (import "future-wasi" "write" (func $write (param anyref i32 i32) (result i32)))
 (global $max_int64 i64 (i64.const 9223372036854775807))
 (global $max_int64_mut (mut i64) (i64.const 9223372036854775807))
 (global $test_float64_bits f64 (f64.const 1.23))
 (memory $0 1)
 (data (i32.const 0) "hello")
 (data "world")
 (table $table 1 1 funcref)
 (elem $elem (i32.const 0) $adder)
 (export "adder" (func $adder))
 (export "memory" (memory $0))
 (export "hello" (func $hello))
 (start $start)
 (func $adder (param $0 i32) (param $1 i32) (result i32)
  (block $add (result i32)
   (if
    (i32.const 0)
    (unreachable)
   )
   (i32.add
    (select
     (local.get $0)
     (i32.load8_s
      (local.get $1)
     )
     (i32.const 1)
    )
    (local.get $1)
   )
  )
 )
 (func $start
  (drop
   (call_indirect $table (type $i32_i32_=>_i32)
    (i32.const 3)
    (i32.const 5)
    (i32.const 0)
   )
  )
 )
 (func $hello (param $0 anyref) (result i32)
  (call $write
   (local.get $0)
   (i32.const 0)
   (i32.const 1)
  )
 )
)
(module
 (type $i32_i32_=>_i32 (func (param i32 i32) (result i32)))
 (type $none_=>_none (func))
 (type $anyref_i32_i32_=>_i32 (func (param anyref i32 i32) (result i32)))
 (type $anyref_=>_i32 (func (param anyref) (result i32)))
 (import "future-wasi" "write" (func $write (param anyref i32 i32) (result i32)))
 (memory $0 1)
 (data (i32.const 0) "hello")
 (table $table 1 1 funcref)
 (elem $elem (i32.const 0) $adder)
 (export "adder" (func $adder))
 (export "memory" (memory $0))
 (export "hello" (func $hello))
 (start $start)
 (func $adder (; has Stack IR ;) (param $0 i32) (param $1 i32) (result i32)
  (i32.add
   (select
    (local.get $0)
    (i32.load8_s
     (local.get $1)
    )
    (i32.const 1)
   )
   (local.get $1)
  )
 )
 (func $start (; has Stack IR ;)
  (drop
   (call $adder
    (i32.const 3)
    (i32.const 5)
   )
  )
 )
 (func $hello (; has Stack IR ;) (param $0 anyref) (result i32)
  (call $write
   (local.get $0)
   (i32.const 0)
   (i32.const 1)
  )
 )
)
(module
 (type $type$0 (func (param anyref i32 i32) (result i32)))
 (type $type$1 (func (param i32 i32) (result i32)))
 (type $type$2 (func))
 (type $type$3 (func (param anyref) (result i32)))
 (import "future-wasi" "write" (func $fimport$0 (param anyref i32 i32) (result i32)))
 (memory $0 1)
 (data (i32.const 0) "hello")
 (table $0 1 1 funcref)
 (elem (i32.const 0) $0)
 (export "adder" (func $0))
 (export "memory" (memory $0))
 (export "hello" (func $2))
 (start $1)
 (func $0 (param $0 i32) (param $1 i32) (result i32)
  (i32.add
   (select
    (local.get $0)
    (i32.load8_s
     (local.get $1)
    )
    (i32.const 1)
   )
   (local.get $1)
  )
 )
 (func $1
  (drop
   (call $0
    (i32.const 3)
    (i32.const 5)
   )
  )
 )
 (func $2 (param $0 anyref) (result i32)
  (call $fimport$0
   (local.get $0)
   (i32.const 0)
   (i32.const 1)
  )
 )
)
(module
 (type $type$0 (func (param anyref i32 i32) (result i32)))
 (type $type$1 (func (param i32 i32) (result i32)))
 (type $type$2 (func))
 (type $type$3 (func (param anyref) (result i32)))
 (import "future-wasi" "write" (func $fimport$0 (param anyref i32 i32) (result i32)))
 (memory $0 1)
 (data (i32.const 0) "hello")
 (table $0 1 1 funcref)
 (elem (i32.const 0) $0)
 (export "adder" (func $0))
 (export "memory" (memory $0))
 (export "hello" (func $2))
 (start $1)
 (func $0 (param $0 i32) (param $1 i32) (result i32)
  local.get $0
  local.get $1
  i32.load8_s
  i32.const 1
  select
  local.get $1
  i32.add
 )
 (func $1
  i32.const 3
  i32.const 5
  call $0
  drop
 )
 (func $2 (param $0 anyref) (result i32)
  local.get $0
  i32.const 0
  i32.const 1
  call $fimport$0
 )
)
